### PR TITLE
Support Arbitrary Number of Block Arguments

### PR DESCRIPTION
**Pull request by daumayr**
Expand Blocks so an arbitrary number of arguments can be accepted: `aBlock valueWithArguments: arg` where `arg` is an `Array` containing the arguments.

- change semantics for blocks, so that they throw an error when the number of arguments does not match
- added `#cull*` methods to allow blocks to ignore arguments
- remove the `BlockN` classes, which originated from SOM

This PR solves #94 and #216.

### CODE REVIEW

File: core-lib/Kernel.ns [hunk 1]
```diff
@@ -630,18 +630,21 @@ class Kernel vmMirror: vmMirror = Object <: Value (
...
 
   protected class Block3 = Block ()
   (
     (* Evaluating *)
-    public value                  = ( self value: nil with: nil )
-    public value: arg             = ( self value: arg with: nil )
     public value: arg1 with: arg2 = ( ^ vmMirror blockValue: self with: arg1 with: arg2 )
   )
 
+  protected class BlockArbitrary = Block ()
```
Comment (smarr) [inline core-lib/Kernel.ns]: Do we at this point still need the other block classes?
Couldn't we just do away with Block1-3?

Comment (smarr) [reply core-lib/Kernel.ns]: @richard-roberts any opinions/requirements?
How does Grace handle block activations when the number of arguments doesn't match?

File: core-lib/Kernel.ns [hunk 2]
```diff
@@ -64,7 +64,7 @@ class Kernel vmMirror: vmMirror = Object <: Value (
     public notNil = ( ^ true )
     public ifNil: aBlock = ( ^ self )
     public ifNotNil: aBlock = ( ^ aBlock value )
-    public ifNil: noGoBlock ifNotNil: goBlock = ( ^ goBlock value: self )
```
Comment (smarr) [inline core-lib/Kernel.ns:67]: This was actually correct. But `ifNotNil:` seems wrong.
The reason is that we want to capture the value and have it in the block argument:

```Smalltalk
obj ifNotNil: [:it | it isGuaranteedToExist ].
(* or also *)
obj ifNil: [ #foo ] ifNotNil: [:it <SomeThing> | it baz ].
```

This is used for instance in MiniTest.

Comment (smarr) [reply core-lib/Kernel.ns:67]: More digging gives both uses, with, and without `:it`.
Think we need `cull:` then to give a missing value as `nil`.

Comment (smarr) [reply core-lib/Kernel.ns:67]: Something to consider: http://forum.world.st/cull-protocol-td1560122.html

File: core-lib/Kernel.ns (hunk 1 above)
Comment (ghost) [reply core-lib/Kernel.ns]: We just throw an error; from the Grace spec: "Requesting the apply(...) method evaluates the block; it is an error to provide the wrong number of arguments."

File: core-lib/Kernel.ns (hunk 2 above)
Comment (daumayr) [reply core-lib/Kernel.ns:67]: What is your preference for implementing this: Kernel, Primitive, or both?

Comment (smarr) [reply core-lib/Kernel.ns:67]: First in Kernel, and if it turns out to be a performance concern, we can optimize it?

Comment (daumayr) [reply core-lib/Kernel.ns:67]: Sure, we need a numArgs primitive for that anyway, do we want to expose that for other use-cases or just culling?

Comment (smarr) [reply core-lib/Kernel.ns:67]: Good question. We do not expose it in Symbol either, and I'd consider it reflective information.
So strictly speaking it should be only exposed via mirrors.
Think the best would be to simply use a `vmMirror` primitive directly in the `cull*` implementations.

**Comment by smarr** [comment]
I did a few more cleanups, updated the PR description, and added `#cull*` support.

**Pull request merged**